src/tools/tilesetProcessing/TilesetSplatUpgrader.ts

```diff
@@ -0,0 +1,293 @@
+   * or when the output already exists and `overwrite` was `false`.
+   */
+  async upgrade(
+    tilesetSourceName: string,
+    tilesetTargetName: string,
+    overwrite: boolean
+  ): Promise<void> {
+    const tilesetSourceJsonFileName =
+      Tilesets.determineTilesetJsonFileName(tilesetSourceName);
+    const tilesetTargetJsonFileName =
+      Tilesets.determineTilesetJsonFileName(tilesetTargetName);
+    const tilesetSource = await TilesetSources.createAndOpen(tilesetSourceName);
+    const tilesetTarget = await TilesetTargets.createAndBegin(
+      tilesetTargetName,
+      overwrite
+    );
+
+    await this.upgradeData(
+      tilesetSource,
+      tilesetSourceJsonFileName,
+      tilesetTarget,
+      tilesetTargetJsonFileName
+    );
+
+    await tilesetSource.close();
+    await tilesetTarget.end();
+  }
+
+  /**
+   * Upgrade the specified source tileset, and write it to the given
+   * target.
+   *
+   * The caller is responsible for calling `open` on the given
+   * source and `begin` on the given target before calling this
+   * method, and `close` on the source and `end` on the target
+   * after calling this method.
+   *
+   * @param tilesetSource - The tileset source
+   * @param tilesetSourceJsonFileName - The name of the top-level
+   * tileset JSON file in the source
+   * @param tilesetTarget - The tileset target
+   * @param tilesetTargetJsonFileName - The name of the top-level
+   * tileset JSON file in the target
+   * @returns A promise that resolves when the process is finished
+   * @throws TilesetError When the input could not be processed,
+   * or when the output already exists and `overwrite` was `false`.
+   */
+  async upgradeData(
+    tilesetSource: TilesetSource,
+    tilesetSourceJsonFileName: string,
+    tilesetTarget: TilesetTarget,
+    tilesetTargetJsonFileName: string
+  ): Promise<void> {
+    this.tilesetProcessor = new BasicTilesetProcessor(true);
+    await this.tilesetProcessor.beginData(
+      tilesetSource,
+      tilesetSourceJsonFileName,
+      tilesetTarget,
+      tilesetTargetJsonFileName
+    );
+
+    // Perform the upgrade for the actual tileset object
+    await this.tilesetProcessor.forTileset(
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      async (tileset: Tileset, schema: Schema | undefined) => {
+        await this.upgradeTileset(tileset);
+        return tileset;
+      }
+    );
+
+    // Perform the updates for the tile contents
+    await this.performContentUpgrades(this.tilesetProcessor);
+    await this.tilesetProcessor.end(false);
+  }
+
+  /**
+   * Perform the upgrade of the `Tileset` object, in place.
+   *
+   * @param tileset - The `Tileset` object
+   */
+  async upgradeTileset(tileset: Tileset) {
+    const upgradeRequired = await this.verifyUpgradeRequired(tileset);
+    if (upgradeRequired) {
+      logger.debug(`Upgrading tileset.json`);
+
+      if (typeof tileset.extensions === "undefined") {
```

Comment:
Is there is a CesiumJS-style `defined` in
`import { defined } from "../../base";`

In many cases, I've opted for `if (!tileset.extensions)` for checks like this.

(One could argue back and forth. Yes, `if (!tileset.extensions)` would not trigger when `tileset.extensions = "Fooled ya!"`. But this would already violate the type definition of the `extensions` that was defined in `RootProperty`. I think when the _type_ already dictates that it cannot be one of the `""` or `0` cases that are so problematic for `if`-checks in JavaScript, the `if`-check should be OK...)